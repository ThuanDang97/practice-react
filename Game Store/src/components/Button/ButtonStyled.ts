import { Button, styled } from '@mui/material'
import { IButtonProps } from '@components/Button'

export const ButtonStyled = styled(Button)<IButtonProps>`
  cursor: pointer;
  border: none;
  color: ${({ theme }) => theme.palette.grey[500]};
  &:hover {
    color: ${({ hovercolor, theme }) =>
      hovercolor ? hovercolor : theme.palette.common.white};
  }

  ${({ mbottom }) => `margin-bottom: ${mbottom}px`};
  ${({ mtop }) => `margin-top: ${mtop}px`};
`
