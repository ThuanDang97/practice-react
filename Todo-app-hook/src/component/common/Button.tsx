import { memo, useCallback } from 'react';

type TProps = {
  id: number;
  title: string;
  onUpdateFilterValue: (value: string) => void;
};

const Button = (props: TProps): JSX.Element => {
  const { title, onUpdateFilterValue } = props;

  const handleUpdateFilterValue = useCallback(() => {
    onUpdateFilterValue(title);
  }, [onUpdateFilterValue]);

  return (
    <button className="btn-todo" onClick={handleUpdateFilterValue}>
      {title}
    </button>
  );
};

export default memo(Button);
// export default Button;
