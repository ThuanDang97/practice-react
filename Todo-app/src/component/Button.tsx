import React from 'react';
import { Component } from 'react';

type Props = {
  id: number;
  title: string;
  onUpdateFilterValue: Function;
  isActived: boolean;
};

class Button extends Component<Props> {
  constructor(props: Props) {
    super(props);
  }

  render(): JSX.Element {
    return (
      <>
        <li>
          <a
            className={`${this.props.isActived} ? 'selected' : ''`}
            onClick={() => this.props.onUpdateFilterValue(this.props.title)}
          >
            {this.props.title}
          </a>
        </li>
        <span></span>
      </>
    );
  }
}

export default Button;
