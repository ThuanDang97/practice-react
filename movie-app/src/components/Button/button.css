.btn {
  border: 1px solid transparent;
  color: white;
  height: 50px;
  cursor: pointer;
  text-align: center;
  font-size: 20px;
  font-weight: 700;
}

.btn:hover {
  color: rgba(0, 0, 0, 0.8);
}

.btn-layout {
  width: 100%;
  margin-top: 20px;
}

button[disabled] {
  background-color: #ececec;
  color: rgba(0, 0, 0, 0.5);
  cursor: no-drop;
}

button[disabled]:hover {
  color: rgba(0, 0, 0, 0.5);
}

.primary {
  background-color: rgba(1, 180, 228, 1);
}

.success {
  background-color: rgba(25, 135, 84, 1);
}

.danger {
  background-color: rgba(220, 53, 69, 1);
}

.w-25 {
  width: 25%;
}

.w-50 {
  width: 50%;
}

.w-100 {
  width: 100%;
}

.rd-1 {
  border-radius: 10px;
}

.rd-2 {
  border-radius: 30px;
}

.text {
  border: none;
  background-color: transparent;
  color: rgba(0, 0, 0, 0.8);
}

.contained {
  border: 1px solid transparent;
}

.outline {
  border: 1px solid rgba(0, 0, 0, 0.5);
  background-color: transparent;
  color: rgba(0, 0, 0, 0.8);
}
