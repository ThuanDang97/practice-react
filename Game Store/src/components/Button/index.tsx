import React from 'react'
import { ButtonProps } from '@mui/material'

// Styles
import { ButtonStyled } from '@components/Button/ButtonStyled'

export interface IButtonProps extends ButtonProps {
  children: React.ReactNode
  onClick?: (e: React.MouseEvent<HTMLElement>) => void
  disable?: boolean
  startIcon?: React.ReactNode
  endIcon?: React.ReactNode
  width?: number | string
  height?: number
  type?: 'button' | 'submit' | 'reset'
  hovercolor?: string
  mbottom?: number
  mtop?: number
}

const Button = ({
  children = 'Button',
  disable = false,
  color = 'primary',
  startIcon = false,
  endIcon = false,
  onClick,
  width = '100%',
  height,
  type = 'button',
  hovercolor = 'white',
  ...props
}: IButtonProps) => {
  return (
    <ButtonStyled
      width={width}
      height={height}
      color={color}
      disabled={disable}
      variant="contained"
      startIcon={startIcon ? startIcon : false}
      endIcon={endIcon ? endIcon : false}
      onClick={onClick}
      type={type}
      hovercolor={hovercolor}
      {...props}
    >
      {children}
    </ButtonStyled>
  )
}

export default Button
