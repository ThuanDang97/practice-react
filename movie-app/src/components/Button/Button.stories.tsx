import { Story } from '@storybook/react';
import { Button } from '../Button/Button';
import { TProps } from '../Button/Button';

export default {
  title: 'Components/Button',
  component: Button,
};

const Template: Story<TProps> = (args): JSX.Element => <Button {...args} />;

export const Basic: Story<TProps> = Template.bind({});
Basic.args = {
  children: 'Press me',
  typeButton: 'button',
  radiant: 'contained',
  color: 'primary',
  width: 'w-25',
  radius: 'rd-1',
  isDisable: false,
};

export const Disable: Story<TProps> = Template.bind({});
Disable.args = {
  children: 'Press me',
  typeButton: 'button',
  radiant: 'contained',
  color: 'primary',
  width: 'w-25',
  radius: 'rd-1',
  isDisable: true,
};
