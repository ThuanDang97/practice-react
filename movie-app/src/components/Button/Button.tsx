import { MouseEvent, ReactNode } from 'react';
import './button.css';

export type TProps = {
  children: string | ReactNode;
  radiant: 'text' | 'contained' | 'outline';
  typeButton: 'button' | 'reset' | 'submit';
  color: 'primary' | 'success' | 'danger';
  width: 'w-25' | 'w-50' | 'w-100';
  radius: 'rd-1' | 'rd-2';
  onHandleClick: (e: MouseEvent<HTMLButtonElement>) => void;
  isDisable: boolean;
};

const Button = (props: TProps): JSX.Element => {
  const {
    children,
    typeButton,
    color,
    width,
    radiant,
    radius,
    onHandleClick,
    isDisable,
  } = props;

  return (
    <button
      type={typeButton}
      className={`btn ${color} ${radiant} ${radius} ${width} `}
      onClick={onHandleClick}
      disabled={isDisable}
    >
      {children}
    </button>
  );
};

export { Button };
