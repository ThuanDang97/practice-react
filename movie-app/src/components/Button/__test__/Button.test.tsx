import { screen, render } from '@testing-library/react';
import { setup } from '../../../support/Setup';
import { Button, TProps } from '../Button';
describe('Test Button component', () => {
  const defaultProps: TProps = {
    children: 'Button',
    radiant: 'text',
    typeButton: 'button',
    color: 'primary',
    width: 'w-25',
    radius: 'rd-1',
    onHandleClick: jest.fn(),
    isDisable: false,
  };

  const customProps: TProps = {
    children: 'Button',
    radiant: 'text',
    typeButton: 'button',
    color: 'primary',
    width: 'w-25',
    radius: 'rd-1',
    onHandleClick: jest.fn(),
    isDisable: true,
  };

  test('Component Button matches DOM Snapshot', () => {
    const { container } = render(<Button {...defaultProps}></Button>);

    expect(container).toMatchSnapshot();
  });

  test('Component Button should render correctly', () => {
    setup(<Button {...defaultProps}></Button>);

    expect(screen.getByRole('button')).toHaveTextContent('Button');
  });

  test('Component Button should called handleClick', async () => {
    const { user } = setup(<Button {...defaultProps}></Button>);

    await user.click(screen.getByRole('button'));

    expect(defaultProps.onHandleClick).toHaveBeenCalled();
    expect(defaultProps.onHandleClick).toHaveBeenCalledTimes(1);
  });

  test('Component Button is disable and not called handleClick', async () => {
    const { user } = setup(<Button {...customProps}></Button>);

    await user.click(screen.getByRole('button'));

    expect(screen.getByRole('button')).toBeDisabled();
    expect(customProps.onHandleClick).not.toHaveBeenCalled();
  });
});
