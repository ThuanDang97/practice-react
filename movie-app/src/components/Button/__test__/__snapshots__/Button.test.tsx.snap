// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button component Component Button matches DOM Snapshot 1`] = `
<div>
  <button
    class="btn primary text rd-1 w-25 "
    type="button"
  >
    Button
  </button>
</div>
`;
